torch/nn/utils/clip_grad.py

```diff
@@ -1,15 +1,17 @@
 import warnings
+from typing import Union, Iterable
+
 import torch
 from torch._six import inf
-from typing import Union, Iterable
+from torch.utils._foreach_utils import _group_tensors_by_device_and_dtype
 
 _tensor_or_tensors = Union[torch.Tensor, Iterable[torch.Tensor]]
 
 __all__ = ['clip_grad_norm_', 'clip_grad_norm', 'clip_grad_value_']
 
 def clip_grad_norm_(
         parameters: _tensor_or_tensors, max_norm: float, norm_type: float = 2.0,
-        error_if_nonfinite: bool = False) -> torch.Tensor:
+        error_if_nonfinite: bool = False, foreach: bool = True) -> torch.Tensor:
```

Comment:
So we actually cannot just set it to True, because most foreach ops only have a dispatch defined for CUDA and CPU (and not any other backend like MKLDNN or MPS etc) (see the ops in native_functions.yaml)

Reply:
In the newest commit the behaviour is that if foreach is True, it will run foreach on CPU and CUDA, but still keep running the other device groups like before with a python for loop. That's okay right?

Reply:
Yes, that's right. cc @albanD to confirm we aren't falling subject to BC breaking risks by having foreach to default to True here.

Could you please document the default for the foreach kwarg in both the function documentations?

Reply:
I think one of the benefit to have 3 states (None, True, False) is that the user has an API to force this to be used and they will get a nice error if it doesn't.
I think it is important to have such a nob for users who care about perf.

Reply:
Got it, I implemented the 3-way state and doc changes in https://github.com/pytorch/pytorch/pull/91846/commits/cfea3967a017befad38829e75cbc95fb966261f5